streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,303 @@
+            final UUID clientId = clientStateEntry.getKey();
+            final ClientState clientState = clientStateEntry.getValue();
+
+            clientState.clientTags().forEach((tagKey, tagValue) -> {
+                tagKeyToValues.computeIfAbsent(tagKey, ignored -> new HashSet<>()).add(tagValue);
+                tagEntryToClients.computeIfAbsent(new TagEntry(tagKey, tagValue), ignored -> new HashSet<>()).add(clientId);
+            });
+        }
+    }
+
+    private static int assignStandbyTasksToClientsWithDifferentTags(final ConstrainedPrioritySet standbyTaskClientsByTaskLoad,
+                                                                    final TaskId activeTaskId,
+                                                                    final UUID activeTaskClient,
+                                                                    final Set<String> rackAwareAssignmentTags,
+                                                                    final Map<UUID, ClientState> clientStates,
+                                                                    final Map<TaskId, Integer> tasksToRemainingStandbys,
+                                                                    final Map<String, Set<String>> tagKeyToValues,
+                                                                    final Map<TagEntry, Set<UUID>> tagEntryToClients) {
+        standbyTaskClientsByTaskLoad.offerAll(clientStates.keySet());
+
+        // We set countOfUsedClients as 1 because client where active task is located has to be considered as used.
+        int countOfUsedClients = 1;
+        int numRemainingStandbys = tasksToRemainingStandbys.get(activeTaskId);
+
+        final Map<TagEntry, Set<UUID>> tagEntryToUsedClients = new HashMap<>();
+
+        fillClientsOnAlreadyUsedTagEntries(
+            activeTaskClient,
+            countOfUsedClients,
+            rackAwareAssignmentTags,
+            clientStates,
+            tagEntryToClients,
+            tagKeyToValues,
+            tagEntryToUsedClients
+        );
+
+        while (numRemainingStandbys > 0) {
+            final UUID clientOnUnusedTagDimensions = standbyTaskClientsByTaskLoad.poll(
+                activeTaskId, uuid -> !isClientUsedOnAnyOfTheTagEntries(uuid, tagEntryToUsedClients)
+            );
+
+            if (clientOnUnusedTagDimensions == null) {
+                break;
+            }
+
+            clientStates.get(clientOnUnusedTagDimensions).assignStandby(activeTaskId);
+
+            countOfUsedClients++;
+            numRemainingStandbys--;
+
+            if (numRemainingStandbys == 0) {
+                break;
+            }
+
+            fillClientsOnAlreadyUsedTagEntries(
+                clientOnUnusedTagDimensions,
+                countOfUsedClients,
+                rackAwareAssignmentTags,
+                clientStates,
+                tagEntryToClients,
+                tagKeyToValues,
+                tagEntryToUsedClients
+            );
+        }
```

Comment:
Wouldn't this be equivalent and maybe a bit more concise?

```suggestion
        UUID lastUsedclient = activeTaskClient;
        do {
            fillClientsOnAlreadyUsedTagEntries(
                lastUsedclient,
                countOfUsedClients,
                rackAwareAssignmentTags,
                clientStates,
                tagEntryToClients,
                tagKeyToValues,
                tagEntryToUsedClients
            );

            final UUID clientOnUnusedTagDimensions = standbyTaskClientsByTaskLoad.poll(
                activeTaskId, uuid -> !isClientUsedOnAnyOfTheTagEntries(uuid, tagEntryToUsedClients)
            );

            if (clientOnUnusedTagDimensions == null) {
                break;
            }

            clientStates.get(clientOnUnusedTagDimensions).assignStandby(activeTaskId);

            countOfUsedClients++;
            numRemainingStandbys--;

            lastUsedclient = clientOnUnusedTagDimensions;
        } while (numRemainingStandbys > 0);
```

Reply:
Good suggestion!